### PR TITLE
UCP: Initial imlementation of server listener object.

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5720/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5724/ for details (Mellanox internal link).

File: src/ucp/core/ucp_listener.c [hunk 1]
```diff
@@ -0,0 +1,156 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2001-2015.  ALL RIGHTS RESERVED.
```
Comment (miked-mellanox) [inline src/ucp/core/ucp_listener.c]: copyright year

File: src/ucp/core/ucp_listener.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2001-2015.  ALL RIGHTS RESERVED.
```
Comment (miked-mellanox) [inline src/ucp/core/ucp_listener.h]: year

File: src/ucp/core/ucp_context.h [hunk 3]
```diff
@@ -209,6 +209,19 @@ typedef struct ucp_am_handler {
     }
 
 
+#define UCP_CHECK_PARAM_NON_NULL(_param, _status, _action) \
+    if ((_param) == NULL) { \
+        ucs_error("The parameter %s must not be NULL", #_param); \
```
Comment (alinask) [inline src/ucp/core/ucp_context.h]: should the error messages start with small letters? (same for all error msgs in this pr)

File: src/ucp/core/ucp_listener.c [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        }
+
+        if (params->field_mask & UCP_WORKER_LISTENER_PARAM_FIELD_CALLBACK) {
+            UCP_CHECK_PARAM_NON_NULL(params->ep_accept_handler.cb, status, goto err_free);
+            listener->cb  = params->ep_accept_handler.cb;
+            listener->arg = params->ep_accept_handler.arg;
+        } else {
+            /* TODO if callback is not specified, avoid creating endpoint on server side
+             * if not really needed (not p2p tl)
+             */
```
Comment (alinask) [inline src/ucp/core/ucp_listener.c]: we are going to create an ep anyway right? for sending a conn_req to the client.

File: src/ucp/core/ucp_context.h (hunk 3 above)
Comment (yosefe) [reply src/ucp/core/ucp_context.h]: yes

File: src/ucp/core/ucp_listener.c (hunk 4 above)
Comment (yosefe) [reply src/ucp/core/ucp_listener.c]: right, will remove the comment

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5728/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5731/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@alinask @miked-mellanox fixed

File: src/ucp/core/ucp_listener.c [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    ucp_listener_accept_t *accept;
+    ucs_status_t status;
+    int prog_id;
+
+    ucs_trace("listener %p: got connection request", listener);
+
+    /* if user provided a callback for accepting new connection, launch it on
+     * the main thread
+     */
+    if (listener->cb != NULL) {
```
Comment (alinask) [inline src/ucp/core/ucp_listener.c]: won't this always be true now that the cb is set to ucs_empty_function?

**Comment by alinask** [comment]
:+1:

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5733/ for details (Mellanox internal link).

**Pull request merged**